File: Kernel Simulator/Misc/Scripts/BeepSynth.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.IO

Public Module BeepSynth

    ''' <summary>
    ''' Probes the synth file and plays it back using the console PC speaker
    ''' </summary>
    ''' <param name="file">A file name in current shell path</param>
    ''' <returns>True if successful, False if unsuccessful.</returns>
    ''' <exception cref="EventsAndExceptions.InvalidSynthException"></exception>
    ''' <exception cref="FileNotFoundException"></exception>
    Public Function ProbeSynth(ByVal file As String) As Boolean
        file = NeutralizePath(file)
        Wdbg("I", "Probing {0}...", file)
        If IO.File.Exists(file) Then
            Dim FStream As New StreamReader(file)
            Wdbg("I", "Opened StreamReader(file) with the length of {0}", FStream.BaseStream.Length)
            If FStream.ReadLine = "KS-BSynth" Then
                'Comments are ignored in the file. Comment format: - <message>
                Wdbg("I", "File is scripted")
                While Not FStream.EndOfStream
                    Dim line As String = FStream.ReadLine
                    Wdbg("I", "Line: {0}", line)
                    If Not line.StartsWith("-") And Not line = "" Then
                        Try
                            Wdbg("I", "Not a comment. Getting frequency and time...")
                            Dim freq As Integer = line.Remove(line.IndexOf(","))
                            Dim ms As Integer = line.Substring(line.IndexOf(",") + 1)
                            Wdbg("I", "Got frequency {0} Hz and time {1} ms", freq, ms)
                            Console.Beep(freq, ms)
                        Catch ex As Exception
                            Wdbg("E", "Not a comment and not a synth line. ({0})", line)
                            Throw New EventsAndExceptions.InvalidSynthException(DoTranslation("Failed to probe a synth line.", currentLang))
                        End Try
                    End If
                End While
                Return True
            Else
                Wdbg("E", "File is not scripted")
                Throw New EventsAndExceptions.InvalidSynthException(DoTranslation("The file isn't a scripted synth file.", currentLang))
            End If
        Else
            Wdbg("E", "File doesn't exist")
            Throw New FileNotFoundException(DoTranslation("Scripted file {0} does not exist.", currentLang).FormatString(file))
        End If
        Return False
    End Function

End Module

File: public/Nitrocid.Addons/Nitrocid.Extras.BeepSynth/Commands/BeepSynth.cs
//
// Nitrocid KS  Copyright (C) 2018-2025  Aptivi
//
// This file is part of Nitrocid KS
//
// Nitrocid KS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Nitrocid KS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using Nitrocid.ConsoleBase.Colors;
using Nitrocid.ConsoleBase.Writers;
using Nitrocid.Extras.BeepSynth.Tools;
using Nitrocid.Files;
using Nitrocid.Files.Operations.Querying;
using Nitrocid.Kernel.Exceptions;
using Nitrocid.Kernel.Threading;
using Nitrocid.Languages;
using Nitrocid.Shell.ShellBase.Commands;
using Terminaux.Base;

namespace Nitrocid.Extras.BeepSynth.Commands
{
    /// <summary>
    /// Plays a beep synth
    /// </summary>
    /// <remarks>
    /// This command allows you to read a beep synth file represented with the help of JSON.
    /// </remarks>
    class BeepSynthCommand : BaseCommand, ICommand
    {
        public override int Execute(CommandParameters parameters, ref string variableValue)
        {
            string path = FilesystemTools.NeutralizePath(parameters.ArgumentsList[0]);
            if (!Checking.FileExists(path))
            {
                TextWriters.Write(Translate.DoTranslation("Beep synth file doesn't exist."), KernelColorType.Error);
                return KernelExceptionTools.GetErrorCode(KernelExceptionType.Console);
            }    
            var synthInfo = SynthTools.GetSynthInfoFromFile(path);
            TextWriters.Write(Translate.DoTranslation("Now playing") + ": ", false, KernelColorType.ListEntry);
            TextWriters.Write(synthInfo.Name, KernelColorType.ListValue);
            for (int i = 0; i < synthInfo.Chapters.Length; i++)
            {
                SynthInfo.Chapter chapter = synthInfo.Chapters[i];
                TextWriters.Write($"- [{i + 1}/{synthInfo.Chapters.Length}] ", false, KernelColorType.NeutralText);
                TextWriters.Write(Translate.DoTranslation("Chapter name") + ": ", false, KernelColorType.ListEntry);
                TextWriters.Write(chapter.Name, KernelColorType.ListValue);
                for (int j = 0; j < chapter.Synths.Length; j++)
                {
                    string synth = chapter.Synths[j];
                    var split = synth.Split(' ');
                    if (split.Length != 2)
                    {
                        TextWriters.Write(Translate.DoTranslation("Synth representation is invalid.") + $" [{i + 1}.{j + 1}]", KernelColorType.Error);
                        return KernelExceptionTools.GetErrorCode(KernelExceptionType.Console);
                    }
                    if (!int.TryParse(split[0], out int freq))
                    {
                        TextWriters.Write(Translate.DoTranslation("Frequency is invalid.") + $" [{i + 1}.{j + 1}]", KernelColorType.Error);
                        return KernelExceptionTools.GetErrorCode(KernelExceptionType.Console);
                    }
                    if (!int.TryParse(split[1], out int ms))
                    {
                        TextWriters.Write(Translate.DoTranslation("Duration is invalid.") + $" [{i + 1}.{j + 1}]", KernelColorType.Error);
                        return KernelExceptionTools.GetErrorCode(KernelExceptionType.Console);
                    }
                    if (freq == 0)
                        ThreadManager.SleepNoBlock(ms);
                    else
                        ConsoleWrapper.Beep(freq, ms);
                }
            }
            return 0;
        }

    }
}
